unable to build with jbuilder

Does this actually compile for anyone else?

I'm getting *tons* of errors like this:
```
File "lib/endpoint.ml", line 22, characters 0-50:
Warning 32: unused value $.
File "lib/endpoint.ml", line 36, characters 6-9:
Warning 32: unused value map.
File "_none_", line 1:
Warning 32: unused value sizeof_ring_shared.
File "_none_", line 1:
Warning 32: unused value hexdump_ring_shared.
File "_none_", line 1:
Warning 32: unused value copy_vchan_interface_left.
File "_none_", line 1:
Warning 32: unused value set_vchan_interface_left.
File "_none_", line 1:
Warning 32: unused value blit_vchan_interface_left.
File "_none_", line 1:
Warning 32: unused value copy_vchan_interface_right.
File "_none_", line 1:
Warning 32: unused value set_vchan_interface_right.
File "_none_", line 1:
Warning 32: unused value blit_vchan_interface_right.
File "_none_", line 1:
Warning 32: unused value hexdump_vchan_interface.
File "lib/endpoint.ml", line 263, characters 4-14:
Warning 32: unused value data_ready.
File "lib/endpoint.ml", line 275, characters 4-16:
Warning 32: unused value buffer_space.
File "lib/endpoint.ml", line 488, characters 22-23:
Warning 27: unused variable n.
File "lib/endpoint.ml", line 537, characters 4-14:
Warning 32: unused value disconnect.
File "lib/endpoint.ml", line 1:
Error: Some fatal warnings were triggered (15 occurrences)
Makefile:5: recipe for target 'build' failed
```

This can be fixed by inserting `let _ = foobar in` everywhere, or by `_` in record pattern matches, but it'd be easier to just turn off warning 32. I have no idea how to that with jbuilder.

Now `opam` complains. Am I doing something wrong?
```
user@pubdev:~$ opam install vchan
The following actions will be performed:
 - install   vchan.2.3.1
=== 1 to install ===

=-=- Synchronizing package archives -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

=-=- Installing packages =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
Building vchan.2.3.1:
  ./configure --disable-xenctrl --disable-xen
  make build
  make install
  make js-install
[ERROR] The compilation of vchan.2.3.1 failed.
Removing vchan.2.3.1.
  make js-uninstall
  ocamlfind remove vchan
[WARNING] Directory /home/user/.opam/4.04.0/doc/vchan is not empty, not removing


#=== ERROR while installing vchan.2.3.1 =======================================#
# opam-version 1.2.0
# os           linux
# command      make build
# path         /home/user/.opam/4.04.0/build/vchan.2.3.1
# compiler     4.04.0
# exit-code    2
# env-file     /home/user/.opam/4.04.0/build/vchan.2.3.1/vchan-8625-6d07ae.env
# stdout-file  /home/user/.opam/4.04.0/build/vchan.2.3.1/vchan-8625-6d07ae.out
# stderr-file  /home/user/.opam/4.04.0/build/vchan.2.3.1/vchan-8625-6d07ae.err
### stdout ###
# ...[truncated]
# /home/user/.opam/4.04.0/bin/ocamlfind ocamlmklib -o lib/vchan_stubs lib/vchan_stubs.o
# /home/user/.opam/4.04.0/bin/ocamlfind ocamldep -package xenstore.client -package sexplib -package result -package ppx_sexp_conv -package ppx_deriving -package mirage-flow-lwt -package lwt -package io-page -package cstruct.ppx -package cstruct -modules lib/endpoint.mli > lib/endpoint.mli.depends
# + /home/user/.opam/4.04.0/bin/ocamlfind ocamldep -package xenstore.client -package sexplib -package result -package ppx_sexp_conv -package ppx_deriving -package mirage-flow-lwt -package lwt -package io-page -package cstruct.ppx -package cstruct -modules lib/endpoint.mli > lib/endpoint.mli.depends
# sh: 1: /home/user/.opam/4.04.0/lib/ppx_cstruct/./ppx.exe: not found
# File "lib/endpoint.mli", line 1:
# Error: Error while running external preprocessor
# Command line: /home/user/.opam/4.04.0/lib/ppx_cstruct/./ppx.exe --as-ppx '/tmp/camlppx92cf83' '/tmp/camlppx002c9d'
# 
# Command exited with code 2.
# Makefile:19: recipe for target 'build' failed
### stderr ###
# ...[truncated]
# File "configure.ml", line 121, characters 49-65:
# Warning 3: deprecated: String.uppercase
# Use String.uppercase_ascii instead.
# Looking for ocamlfind package xen-evtchn: ok
# Looking for ocamlfind package xen-gnt-unix: missing
# Looking for ocamlfind package mirage-xen: missing
# W: Cannot find source file matching module 'vchan' in library vchan.
# W: Use InterfacePatterns or ImplementationPatterns to define this file with feature "source_patterns".
# E: Failure("Command ''/home/user/.opam/4.04.0/bin/ocamlbuild' lib/libvchan_stubs.a lib/dllvchan_stubs.so lib/vchan.cma lib/vchan.cmxa lib/vchan.a lib/vchan.cmxs -use-ocamlfind -tag debug -j 4' terminated with error code 10")
# make: *** [build] Error 1
```

--- a/lib/endpoint.ml
+++ b/lib/endpoint.ml
@@ -18,9 +18,6 @@ open Sexplib.Std
 open Lwt
 open Result
 
-external (|>) : 'a -> ('a -> 'b) -> 'b = "%revapply";;
-external ( $ ) : ('a -> 'b) -> 'a -> 'b = "%apply"
-
 let ( >>= ) = Lwt.bind
 
 let i_int (i:int) = ignore i
@@ -33,7 +30,6 @@ module Int32 = struct
 end
 
 module Opt = struct
-  let map f o = match o with Some e -> Some (f e) | None -> None
   let iter f o = match o with Some e -> f e | None -> ()
 end
 
@@ -47,6 +43,8 @@ external atomic_fetch_and : Cstruct.buffer -> int -> int -> int = "stub_atomic_f
 
 (* XXX: the xen headers do not use __attribute__(packed). Edit vb: Was
    OK for me. *)
+
+[@@@warning "-32"]
 
 (* matches xen/include/public/io/libxenvchan.h:ring_shared *)
 [%%cstruct
@@ -69,6 +67,8 @@ type vchan_interface = {
   srv_notify: uint8_t;
 } [@@little_endian]
 ]
+
+[@@@warning "+32"]
 
 let get_ro v = get_vchan_interface_right_order v
 let get_lo v = get_vchan_interface_left_order v
@@ -260,10 +260,6 @@ let fast_get_data_ready (vch: t) request =
   if ready >= request then ready else
     (request_notify vch Write; Int32.(rd_prod vch - rd_cons vch |> to_int))
 
-let data_ready (vch: t) =
-  request_notify vch Write;
-  Int32.(rd_prod vch - rd_cons vch |> to_int)
-
 let fast_get_buffer_space (vch: t) request =
   let ready = wr_ring_size vch - Int32.(wr_prod vch - wr_cons vch |> to_int) in
   if ready > request then ready else
@@ -271,10 +267,6 @@ let fast_get_buffer_space (vch: t) request =
       request_notify vch Read;
       wr_ring_size vch - Int32.(wr_prod vch - wr_cons vch |> to_int)
     )
-
-let buffer_space (vch: t) =
-  request_notify vch Read;
-  wr_ring_size vch - Int32.(wr_prod vch - wr_cons vch |> to_int)
 
 let state vch =
   let client_state =
@@ -485,7 +477,7 @@ let client ~domid ~port () =
   let map_locations grants l = match l with
   | Location.Within_shared_page offset ->
     None, Cstruct.sub v (Location.to_offset offset) (Location.to_length l)
-  | Location.External n ->
+  | Location.External _n ->
     let mapping = M.mapv ~grants ~rw:true in
     Some mapping, Io_page.to_cstruct (M.buf_of_mapping mapping) in
   let w_map, w_buf = map_locations lgrants lo in
@@ -533,10 +525,5 @@ let close (vch: t) =
       E.close vch.evtchn;
       return ()
   end
-
-let disconnect (vch: t) =
-  (* In the vchan protocol close doesn't wait for any kind of acknowledgement,
-     so disconnect and close are the same. *)
-  close vch
 
 end

--- a/lib/in_memory.ml
+++ b/lib/in_memory.ml
@@ -28,11 +28,11 @@ module Config = struct
 
   let c = Lwt_condition.create ()
 
-  let write ~client_domid ~port t =
+  let write ~client_domid:_ ~port t =
     Hashtbl.replace tbl port t;
     return ()
 
-  let read ~server_domid ~port =
+  let read ~server_domid:_ ~port =
     let rec loop () =
       if Hashtbl.mem tbl port
       then return (Hashtbl.find tbl port)
@@ -41,7 +41,7 @@ module Config = struct
         loop () in
     loop ()
 
-  let delete ~client_domid ~port =
+  let delete ~client_domid:_ ~port =
     Hashtbl.remove tbl port;
     return ()
 
@@ -81,7 +81,7 @@ module Memory = struct
   let individual_pages = Hashtbl.create 16
   let big_mapping = Hashtbl.create 16
 
-  let share ~domid ~npages ~rw =
+  let share ~domid:_ ~npages ~rw:_ =
     let mapping = Io_page.get npages in
     let grants = get_n npages in
     let share = { grants; mapping } in
@@ -134,7 +134,7 @@ module Memory = struct
     Hashtbl.replace currently_mapped first ();
     { mapping; grants }
 
-  let unmap { mapping; grants } =
+  let unmap { mapping = _; grants } =
     let first = snd (List.hd grants) in
     if Hashtbl.mem currently_mapped first
     then Hashtbl.remove currently_mapped first

--- a/lib/in_memory_events.ml
+++ b/lib/in_memory_events.ml
@@ -20,8 +20,6 @@ open Sexplib.Std
 
 open Lwt
 
-type 'a io = 'a Lwt.t
-
 type port = int [@@deriving sexp_of]
 
 let port_of_string x = `Ok (int_of_string x)

--- a/lib_test/test.ml
+++ b/lib_test/test.ml
@@ -149,6 +149,7 @@ let test_write_wraps () = Lwt_main.run (
       for i = 0 to Cstruct.len ring - 1 do Cstruct.set_char ring i 'X' done;
       V.write server ring >>|= fun () ->
       V.read client >>!= fun buf ->
+      assert_equal ~printer:(fun x -> x) (string_of_cstruct ring) (string_of_cstruct buf);
       (* writing and reading 1 byte will ensure we have consumed the previous chunk
          (read doesn't perform a copy, see ack_up_to) *)
       V.write server (cstruct_of_string "!") >>|= fun () ->

--- a/lwt/vchan_lwt_unix.ml
+++ b/lwt/vchan_lwt_unix.ml
@@ -65,8 +65,6 @@ module IO = struct
 
   let write = Lwt_io.write
 
-  let write_line = Lwt_io.write_line
-
   let flush = Lwt_io.flush
 
 end

--- a/vchan-unix.opam
+++ b/vchan-unix.opam
@@ -1,4 +1,5 @@
 opam-version: "1.2"
+synopsis:      "Xen Vchan implementation"
 maintainer:    "jonathan.ludlam@eu.citrix.com"
 authors: [
   "Vincent Bernardoff"

--- a/vchan-xen.opam
+++ b/vchan-xen.opam
@@ -1,4 +1,5 @@
 opam-version: "1.2"
+synopsis:      "Xen Vchan implementation"
 maintainer:    "jonathan.ludlam@eu.citrix.com"
 authors: [
   "Vincent Bernardoff"

--- a/vchan.opam
+++ b/vchan.opam
@@ -1,4 +1,5 @@
 opam-version: "1.2"
+synopsis:      "Xen Vchan implementation"
 maintainer:    "jonathan.ludlam@eu.citrix.com"
 authors: [
   "Vincent Bernardoff"

